null-escape/src/main.rs

```diff
@@ -0,0 +1,218 @@
+    println!("output3: {:?}", output3);
+    assert_eq!(output3, vec![0x5c, 0x5c, 0x00]);
+}
+
+#[test]
+fn test_transform_4() {
+    // 0x5c5c5c5c30 => 0x5c5c5c5c30
+    let test4 = &[0x5c, 0x5c, 0x5c, 0x5c, 0x30];
+    let mut input4 = std::io::Cursor::new(test4);
+    let mut output4: Vec<u8> = Vec::new();
+    let _ = process_input(&mut input4, &mut output4);
+    println!("output4: {:?}", output4);
+    assert_eq!(output4, vec![0x5c, 0x5c, 0x5c, 0x5c, 0x30]);
+}
+
+#[test]
+fn test_transform_5() {
+    // 0x5c5c5c3030 => 0x5c5c0030
+    let test5 = &[0x5c, 0x5c, 0x5c, 0x30, 0x30];
+    let mut input5 = std::io::Cursor::new(test5);
+    let mut output5: Vec<u8> = Vec::new();
+    let _ = process_input(&mut input5, &mut output5);
+    println!("output5: {:?}", output5);
+    assert_eq!(output5, vec![0x5c, 0x5c, 0x00, 0x30]);
+}
+
+#[test]
+fn test_transform_6() {
+    // 0x5c5c5c5c3030 => 0x5c5c5c5c3030
+    let test6 = &[0x5c, 0x5c, 0x5c, 0x5c, 0x30, 0x30];
+    let mut input6 = std::io::Cursor::new(test6);
+    let mut output6: Vec<u8> = Vec::new();
+    let _ = process_input(&mut input6, &mut output6);
+    println!("output6: {:?}", output6);
+    assert_eq!(output6, vec![0x5c, 0x5c, 0x5c, 0x5c, 0x30, 0x30]);
+}
+
+#[test]
+fn test_transform_7() {
+    // 0x5c5c5c40 => 0x5c5c5c40
+    let test7 = &[0x5c, 0x5c, 0x5c, 0x40];
+    let mut input7 = std::io::Cursor::new(test7);
+    let mut output7: Vec<u8> = Vec::new();
+    let _ = process_input(&mut input7, &mut output7);
+    println!("output7: {:?}", output7);
+    assert_eq!(output7, vec![0x5c, 0x5c, 0x5c, 0x40]);
+}
+
+#[test]
+fn test_transform_8() {
+    // 0x5c5c5c5c40 => 0x5c5c5c5c40
+    let test8 = &[0x5c, 0x5c, 0x5c, 0x5c, 0x40];
+    let mut input8 = std::io::Cursor::new(test8);
+    let mut output8: Vec<u8> = Vec::new();
+    let _ = process_input(&mut input8, &mut output8);
+    println!("output8: {:?}", output8);
+    assert_eq!(output8, vec![0x5c, 0x5c, 0x5c, 0x5c, 0x40]);
+}
+
+#[test]
+fn test_transform_9() {
+    // 0x5c5c5c5c40 => 0x5c5c5c5c40
+    let test8 = &[0x5c, 0x24, 0x5c, 0x5c, 0x40];
+    let mut input8 = std::io::Cursor::new(test8);
+    let mut output8: Vec<u8> = Vec::new();
+    let _ = process_input(&mut input8, &mut output8);
+    println!("output9: {:?}", output8);
+    assert_eq!(output8, vec![0x5c, 0x24, 0x5c, 0x5c, 0x40]);
+}
+
+fn write_byte<W>(byte: u8, writer: &mut W) -> Result<usize, std::io::Error>
+    where W: Write
+{
+    let written_bytes = try!(writer.write(&[byte]));
+    if written_bytes < 1 {
+        Err(std::io::Error::new(ErrorKind::WriteZero, "Could not write byte"))
+    } else {
+        Ok(written_bytes)
+    }
+}
+
+fn process_byte<W: Write>(byte: u8, writer: &mut W, count: &mut u64) -> Result<(), std::io::Error> {
+    if byte == 0x30 {
+        if (*count + 1) % 2 == 0 {
+            // we saw 0 or even number of 0x5c before 0x5c30
+            try!(write_byte(0x00, writer));
+            return Ok(());
```

Comment:
This is my lack of applied rust talking, but why do we `try!` here and below instead of just returning the Result of `write_byte` directly? `process_input` is already matching on `Ok(_)`.